// @nx/vue/src/generators/application/lib/add-vite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addVite = addVite;
exports.addVitest = addVitest;
const devkit_1 = require("@nx/devkit");
const vite_1 = require("@nx/vite");
async function addVite(tree, options) {
    // Set up build target (and test target if using vitest)
    const viteTask = await (0, vite_1.viteConfigurationGenerator)(tree, {
        uiFramework: 'none',
        project: options.projectName,
        newProject: true,
        inSourceTests: options.inSourceTests,
        includeVitest: options.unitTestRunner === 'vitest',
        skipFormat: true,
        testEnvironment: 'jsdom',
        addPlugin: options.addPlugin,
    });
    (0, vite_1.createOrEditViteConfig)(tree, {
        project: options.projectName,
        includeLib: false,
        includeVitest: options.unitTestRunner === 'vitest',
        inSourceTests: options.inSourceTests,
        imports: [`import vue from '@vitejs/plugin-vue'`],
        plugins: ['vue()'],
    }, false);
    // Update build to skip type checking since tsc won't work on .vue files.
    // Need to use vue-tsc instead.
    const projectConfig = (0, devkit_1.readProjectConfiguration)(tree, options.projectName);
    if (projectConfig.targets?.build?.options) {
        projectConfig.targets.build.options.skipTypeCheck = true;
        (0, devkit_1.updateProjectConfiguration)(tree, options.projectName, projectConfig);
    }
    return viteTask;
}
async function addVitest(tree, options) {
    const tasks = [];
    const vitestTask = await (0, vite_1.vitestGenerator)(tree, {
        uiFramework: 'none',
        project: options.projectName,
        coverageProvider: 'v8',
        inSourceTests: options.inSourceTests,
        skipFormat: true,
        testEnvironment: 'jsdom',
        addPlugin: options.addPlugin,
        runtimeTsconfigFileName: 'tsconfig.app.json',
    });
    tasks.push(vitestTask);
    (0, vite_1.createOrEditViteConfig)(tree, {
        project: options.projectName,
        includeLib: false,
        includeVitest: true,
        inSourceTests: options.inSourceTests,
        imports: [`import vue from '@vitejs/plugin-vue'`],
        plugins: ['vue()'],
    }, true);
    return tasks;
}
