// @nx/vue/src/generators/application/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = normalizeOptions;
const project_name_and_root_utils_1 = require("@nx/devkit/src/generators/project-name-and-root-utils");
const ts_solution_setup_1 = require("@nx/js/src/utils/typescript/ts-solution-setup");
async function normalizeOptions(host, options) {
    await (0, project_name_and_root_utils_1.ensureRootProjectName)(options, 'application');
    const { projectName, projectRoot: appProjectRoot, importPath, } = await (0, project_name_and_root_utils_1.determineProjectNameAndRootOptions)(host, {
        name: options.name,
        projectType: 'application',
        directory: options.directory,
        rootProject: options.rootProject,
    });
    options.rootProject = appProjectRoot === '.';
    const isUsingTsSolutionConfig = (0, ts_solution_setup_1.isUsingTsSolutionSetup)(host);
    const appProjectName = !isUsingTsSolutionConfig || options.name ? projectName : importPath;
    const e2eProjectName = options.rootProject ? 'e2e' : `${appProjectName}-e2e`;
    const e2eProjectRoot = options.rootProject ? 'e2e' : `${appProjectRoot}-e2e`;
    const parsedTags = options.tags
        ? options.tags.split(',').map((s) => s.trim())
        : [];
    const normalized = {
        ...options,
        projectName: appProjectName,
        appProjectRoot,
        importPath,
        e2eProjectName,
        e2eProjectRoot,
        parsedTags,
        isUsingTsSolutionConfig,
        useProjectJson: options.useProjectJson ?? !isUsingTsSolutionConfig,
    };
    normalized.style = options.style ?? 'css';
    normalized.routing = normalized.routing ?? false;
    normalized.unitTestRunner ??= 'vitest';
    normalized.e2eTestRunner = normalized.e2eTestRunner ?? 'playwright';
    normalized.bundler = normalized.bundler ?? 'vite';
    return normalized;
}
