// nx/src/command-line/configure-ai-agents/command-object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.yargsConfigureAiAgentsCommand = void 0;
const shared_options_1 = require("../yargs-utils/shared-options");
exports.yargsConfigureAiAgentsCommand = {
    command: 'configure-ai-agents',
    describe: 'Configure and update AI agent configurations for your workspace.',
    builder: (yargs) => (0, shared_options_1.withVerbose)(yargs)
        .option('agents', {
        type: 'array',
        string: true,
        description: 'List of AI agents to set up.',
        choices: ['claude', 'codex', 'copilot', 'cursor', 'gemini'],
    })
        .option('interactive', {
        type: 'boolean',
        description: 'When false disables interactive input prompts for options.',
        default: true,
    })
        .option('check', {
        type: 'boolean',
        description: 'Check if any configured agents are out of date and need to be updated. Does not make any changes.',
        default: false,
    })
        .example('$0 configure-ai-agents', 'Interactively select AI agents to update and configure')
        .example('$0 configure-ai-agents --agents claude gemini', 'Prompts for updates and and configuration of Claude and Gemini AI agents')
        .example('$0 configure-ai-agents --check', 'Checks if any configured agents are out of date and need to be updated')
        .example('$0 configure-ai-agents --agents claude gemini --no-interactive', 'Configures and updates Claude and Gemini AI agents without prompts'),
    handler: async (args) => {
        await (await Promise.resolve().then(() => require('./configure-ai-agents'))).configureAiAgentsHandler(args);
    },
};
