// @nx-extend/docusaurus/src/executors/browser/executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = runExecutor;
const tslib_1 = require("tslib");
const lib_1 = require("@docusaurus/core/lib");
const path = tslib_1.__importStar(require("path"));
const path_1 = require("path");
function runExecutor(options, context) {
    return tslib_1.__asyncGenerator(this, arguments, function* runExecutor_1() {
        var _a;
        const projectRoot = path.join(context.root, context.projectsConfigurations.projects[(_a = context.projectName) !== null && _a !== void 0 ? _a : ''].root);
        try {
            yield tslib_1.__await((0, lib_1.build)(projectRoot, {
                bundleAnalyzer: options.bundleAnalyzer,
                outDir: (0, path_1.join)(context.root, options.outputPath),
                minify: options.minify
            }));
            yield yield tslib_1.__await({
                success: true
            });
        }
        catch (err) {
            console.error(err);
            yield yield tslib_1.__await({
                success: false
            });
        }
    });
}
