// @nx-extend/docusaurus/src/executors/dev-server/executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = runExecutor;
const tslib_1 = require("tslib");
const lib_1 = require("@docusaurus/core/lib");
const path = tslib_1.__importStar(require("path"));
function runExecutor(options, context) {
    return tslib_1.__asyncGenerator(this, arguments, function* runExecutor_1() {
        var _a;
        const projectRoot = path.join(context.root, context.projectsConfigurations.projects[(_a = context.projectName) !== null && _a !== void 0 ? _a : ''].root);
        const port = options.port.toString();
        yield tslib_1.__await((0, lib_1.start)(projectRoot, {
            port,
            host: options.host,
            hotOnly: options.hotOnly,
            open: options.open
        }));
        yield yield tslib_1.__await({
            baseUrl: `http://localhost:${port}`,
            success: true
        }
        // This Promise intentionally never resolves, leaving the process running
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        );
        // This Promise intentionally never resolves, leaving the process running
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        yield tslib_1.__await(new Promise(() => { }));
    });
}
